src/Mod/Fem/femmesh/meshtools.py

```diff
@@ -619,87 +619,80 @@ def get_femelement_directions_theshape(femmesh, femelement_table, theshape):
     # print("beam_direction: {}".format(beam_direction))
     # print("defined_angle: {}".format(defined_angle))
 
     import math
-    vector_a = beam_direction
-    angle_rad = (math.pi / 180) * defined_angle
-    nx = abs(math.cos(angle_rad))
-    ny = abs(math.sin(angle_rad))
-    if nx < 0.0000001:
-        nx = 0
-    if ny < 0.0000001:
-        ny = 0
-    # vector_n = [nx, ny]  # not used ATM
-
-    if vector_a[0] != 0:
-        temp_valx = -(vector_a[1] + vector_a[2]) / vector_a[0]
-    else:
-        temp_valx = 0
-    if vector_a[1] != 0:
-        temp_valy = -(vector_a[0] + vector_a[2]) / vector_a[1]
-    else:
-        temp_valy = 0
-    if vector_a[2] != 0:
-        temp_valz = -(vector_a[0] + vector_a[1]) / vector_a[2]
-    else:
-        temp_valz = 0
-
-    # Dot_product_check
-    dot_x = None
-    dot_y = None
-    dot_z = None
-    dot_nt = None
-    if vector_a[0] != 0 and vector_a[1] == 0 and vector_a[2] == 0:
-        normal_n = [temp_valx, nx, ny]
-        dot_x = vector_a[0] * normal_n[0] + vector_a[1] * normal_n[1] + vector_a[2] * normal_n[2]
-    elif vector_a[0] == 0 and vector_a[1] != 0 and vector_a[2] == 0:
-        normal_n = [nx, temp_valy, ny]
-        dot_y = vector_a[0] * normal_n[0] + vector_a[1] * normal_n[1] + vector_a[2] * normal_n[2]
-    elif vector_a[0] == 0 and vector_a[1] == 0 and vector_a[2] != 0:
-        normal_n = [nx, ny, temp_valz]
-        dot_z = vector_a[0] * normal_n[0] + vector_a[1] * normal_n[1] + vector_a[2] * normal_n[2]
-    elif vector_a[0] == 0 and vector_a[1] != 0 and vector_a[2] != 0:
-        normal_n = [nx, temp_valy, ny]
-        dot_y = vector_a[0] * normal_n[0] + vector_a[1] * normal_n[1] + vector_a[2] * normal_n[2]
-    elif vector_a[0] != 0 and vector_a[1] == 0 and vector_a[2] != 0:
-        normal_n = [nx, ny, temp_valz]
-        dot_z = vector_a[0] * normal_n[0] + vector_a[1] * normal_n[1] + vector_a[2] * normal_n[2]
-    else:
-        normal_n = [temp_valx, nx, ny]
-        dot_nt = vector_a[0] * normal_n[0] + vector_a[1] * normal_n[1] + vector_a[2] * normal_n[2]
 
-    dot = vector_a[0] * normal_n[0] + vector_a[1] * normal_n[1] + vector_a[2] * normal_n[2]
-    FreeCAD.Console.PrintLog("{}\n".format(dot))
-    FreeCAD.Console.PrintLog("{}\n".format(normal_n))
+    def normalize(vec: FreeCAD.Vector) -> FreeCAD.Vector:
+        return vec / vec.Length
 
-    # dummy usage of the axis dot to get flake8 quiet
-    del dot_x, dot_y, dot_z, dot, dot_nt
+    def get_normal(vec: FreeCAD.Vector) -> FreeCAD.Vector:
+        x, y, z = vec
+
+        if z == 0:  # in xy plane
+            n = (0, 0, -1)  # default in CalculiX
+        elif z != 0 and x == y == 0:  # vertical beam
+            n = (-1, 0, 0)  # or (0, -1, 0)
+        else:
+            # we want the normal to point downwards
+            n = (x, y, -(x**2 + y**2) / z)
+
+        n = normalize(FreeCAD.Vector(n))
+
+        # sanity check
+        dot = vec.dot(n)
+        if not math.isclose(dot, 0):
```

Comment:
Evaluated to `False` with `dot` of `1e-18` order of magnitude. Apparently, done the comparison wrong - should've used `abs_tol` parameter (see [stackoverflow](https://stackoverflow.com/a/35325039))
Did a fuzzy testing with vector components in range `[-100_000_000, 100_000_000]` (both floats and ints) and the tests passed with `1e-15` absolute tolerance. Would like to set it to `1e-13` - seems quite acceptable

Reply:
Set it to `1e-13`